common/src/jni/main/cpp/NativeCrypto.cpp

```diff
@@ -7316,16 +7316,82 @@ static jint NativeCrypto_SSL_read(JNIEnv* env, jclass, jlong ssl_address, jobjec
         return 0;
     }
 
     OpenSslError sslError;
-    int ret = sslRead(env, ssl, fdObject, shc, reinterpret_cast<char*>(bytes.get() + offset), len,
+    int ret;
+    if (JniUtil::isGetByteArrayElementsLikelyToReturnACopy(array_size)) {
+        if (len <= 1024) {
+            // Allocate small buffers on the stack for performance.
+            jbyte buf[1024];
+            ret = sslRead(env, ssl, fdObject, shc, reinterpret_cast<char*>(&buf[0]), len, sslError,
+                          read_timeout_millis);
+            if (ret > 0) {
+                // Don't bother applying changes if issues were encountered.
+                env->SetByteArrayRegion(b, offset, ret, &buf[0]);
+            }
+        } else {
+            // Allocate larger buffers on the heap.
+            // TODO: Enforce that len >= 0? If not, the cast below could result in a memory
+            // allocation error/attack.
+            jint remaining = len;
+            jint buf_size = (remaining >= 65536) ? 65536 : remaining;
+            std::unique_ptr<jbyte[]> buf(new jbyte[static_cast<unsigned int>(buf_size)]);
+            // TODO: Doesn't the following condition require new(std::nothrow)?
+            if (buf.get() == nullptr) {
+                Errors::jniThrowOutOfMemory(env, "Unable to allocate chunk buffer");
+                return 0;
+            }
+            jint expected_reads = len / buf_size;
+            if (expected_reads > 0) {
+                // TODO: Fix cumulative read timeout? This currently ensures that the cumulative
+                // timeout does not exceed the given timeout, but it is overly strict in that it
+                // allocates uniformly-sized time slots for each call.
+                read_timeout_millis = read_timeout_millis / expected_reads;
+            }
+            ret = 0;
+            while (remaining > 0) {
+                jint temp_ret;
+                jint chunk_size = (remaining >= buf_size) ? buf_size : remaining;
+                // TODO: Make sure it's safe to reuse same callbacks repeatedly.
+                temp_ret = sslRead(env, ssl, fdObject, shc, reinterpret_cast<char*>(buf.get()),
+                                   chunk_size, sslError, read_timeout_millis);
+                if (temp_ret == THROW_SSLEXCEPTION || temp_ret == THROW_SOCKETTIMEOUTEXCEPTION ||
+                    temp_ret == THROWN_EXCEPTION) {
+                    // An error was encountered. Handle below.
+                    // TODO: Is it sufficient to check for negative (or even non-positive) values? 0
+                    // itself appears to be an error code.
+                    ret = temp_ret;
+                    break;
+                }
+                env->SetByteArrayRegion(b, offset, temp_ret, buf.get());
+                // Accumulate bytes read.
+                ret += temp_ret;
+                offset += chunk_size;
+                remaining -= chunk_size;
```

Comment:
Yes, they should be. I changed that locally but forgot to commit.

The same goes for `SSL_write`, but as it turns out, `sslWrite` currently always returns the requested count or an error code